packages/core/index.test-d.ts

```diff
@@ -0,0 +1,50 @@
+import { expectType } from 'tsd'
+import middy from '.'
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'
+
+async function baseHandler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
+  return {
+    statusCode: 200,
+    body: `Hello from ${event.path}`
+  }
+}
+
+type Handler = middy.Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>
+
+// initialize
+let handler = middy(baseHandler)
+expectType<Handler>(handler)
+
+// use with 1 middleware
+handler = handler.use({
+  before: (handler: Handler) => {
+    console.log('Before', handler)
+  },
+  after: (handler: Handler) => {
+    console.log('After', handler)
+  },
+  onError: (handler: Handler) => {
+    console.log('OnError', handler)
+  },
+})
+expectType<Handler>(handler)
+
+
+// use with array of middlewares
+handler = handler.use([{
+  before: (handler: Handler) => {
+    console.log('Before', handler)
+  },
+  after: (handler: Handler) => {
+    console.log('After', handler)
+  },
+  onError: (handler: Handler) => {
+    console.log('OnError', handler)
+  },
+}])
+expectType<Handler>(handler)
+
+// before
+handler.before(function (handler: Handler) {
```

Comment:
This one currently fails for some mismatch of types... got stuck here trying to figure out why...

```
Argument of type '(handler: Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>) => void' is not assignable to parameter of type 'Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>'.
  Type '(handler: Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>) => void' is not assignable to type 'Handler<APIGatewayProxyEvent, APIGatewayProxyResult>'.
    Types of parameters 'handler' and 'event' are incompatible.
      Type 'APIGatewayProxyEvent' is not assignable to type 'Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>'.
        Type 'APIGatewayProxyEventBase<{ [name: string]: any; }>' is not assignable to type 'Handler<APIGatewayProxyEvent, APIGatewayProxyResult>'.
          Type 'APIGatewayProxyEventBase<{ [name: string]: any; }>' provides no match for the signature '(event: APIGatewayProxyEvent, context: Context, callback: Callback<APIGatewayProxyResult>): void | Promise<APIGatewayProxyResult>'.
```

That message sounds a bit cryptic to me right now...

Reply:
I figured this one out. I did not account that we need a `MiddlewareAttachFn` and a `MiddlewareFn` as 2 distinct concepts! :D